Additional tests and cleanup for cache/contenthash

aaronlehmann opened the pull request:
This adds a little extra testing around ** patterns, and adds a
(currently skipped) test for copying directories under symlinks (#2300).

It removes an extra call to `filepath.FromSlash` in `shouldIncludePath`
and an unused argument to that function.

aaronlehmann commented on cache/contenthash/checksum.go:
```diff
@@ -619,13 +619,12 @@ func (cc *cacheContext) includedPaths(ctx context.Context, m *mount, p string, o
 	excludePatternMatcher *fileutils.PatternMatcher,
 ) (bool, error) {
 	if includePatternMatcher != nil {
-		m, err := includePatternMatcher.Matches(filepath.FromSlash(candidate))
```
`FromSlash` is not needed here because `Matches` calls it on its argument.

tonistiigi commented on cache/contenthash/checksum_test.go:
```diff
@@ -554,6 +558,108 @@ func TestChecksumIncludeExclude(t *testing.T) {
+	dgst, err = cc.Checksum(context.TODO(), ref, "prefix/a", ChecksumOpts{IncludePatterns: []string{"**/foo/**", "**/report"}}, nil)
+	require.NoError(t, err)
+	// Now there is a file included
+	require.NotEqual(t, digest.FromBytes([]byte{}), dgst)
```
Would it make sense to add constant expected values to at least some of these to detect regressions?

aaronlehmann replied:
Done.

aaronlehmann commented:
Added a fix for a bug that I discovered incidentally while I was working on #2300 (but this does not fix the symlink issue discussed in #2300). Include patterns like "foo" were not working, because these got transformed to "/foo" for matching, which was wrong and didn't match the copy logic. I've changed it to behave the same way the copy does (matching with the actual patterns), and improved the test coverage.

tonistiigi commented on cache/contenthash/checksum_test.go:
```diff
@@ -460,37 +469,43 @@ func TestChecksumIncludeExclude(t *testing.T) {
+	dgstFoo, err := cc.Checksum(context.TODO(), ref, "", opts(ChecksumOpts{IncludePatterns: []string{"foo"}}), nil)
 	require.NoError(t, err)
-	require.Equal(t, dgstFileData0, dgst)
+	require.NotEqual(t, digest.FromBytes([]byte{}), dgstFoo)
```
Not sure why the equality check is gone here.

aaronlehmann replied:
The old `require.Equal` was checking the wrong value (`dgst` instead of `dgstFoo`). It turns out `dgstFileData0` was never being returned in this case. There are two reasons:

- `includedPaths` returns both `/` and `/foo`, so we don't take the hash of `/foo` as the hash to return. I think we could avoid returning `/` in this case - I don't believe we'd ever use the metadata of `/` in a copy (but please correct me if I'm wrong).
- `Checksum` only returns `includedPaths[0].Record.Digest` when `path.Base(p) == path.Base(includedPaths[0].Path)`, but that's not the case here. `p` is an empty string (copy from root) and `includedPaths[0].Path` is `/foo`.

Do you think the change is good based on this explanation, or should we adjust any of this logic?
